util, core: add spawn_process() helper

tchaikov opened the pull request:
So we can spawn a subprocess with Seastar.

In our use case, Seastar is used to implement a distributed service.
Before the daemon boots and joins the cluster, it calls an external
script to determine the host's location in the network topology and
failure domain settings, and attaches these information in its RPC
request for joining the cluster sent to the cluster management nodes.
The script can be customized by user so he/she can define their own
deployment policy, as long as the script is an executable and outputs
in the format that can be parsed by the daemon calling the script.

This is a convenient way allowing user to customize the daemon's
behavior without introducing yet another RPC service or a plugin
system.

Signed-off-by: Kefu Chai <tchaikov@gmail.com>

tchaikov commented:
still needs to add a runtime test for `pidfd_open()`.

raphaelsc commented:
@tchaikov A few words about the motivation would be nice.

tchaikov commented:
@raphaelsc hi Raphael, just added the use case. if this change makes sense and the API is sane in general, i will add doxygen doc for `process` class and `spawn_process()` function and document them the tutorial also.

as an alternative of `pidfd_open()`, and i will add a periodically nonblocking `waitpid()`.

xemul commented:
Why is Linux 5.3 requirement?

tchaikov commented:
> Why is Linux 5.3 requirement?

@xemul in current implementation, [pidfd_open()](https://man7.org/linux/man-pages/man2/pidfd_open.2.html) is used to obtain a fd referring the child process. but that API was introduced in linux 5.3. see https://man7.org/linux/man-pages/man2/pidfd_open.2.html#VERSIONS

but i plan to add a polling machinery to `waitpid()` in non-blocking mode if `pidfd_open()` is not available.

xemul commented on include/seastar/core/reactor.hh:
```diff
@@ -532,7 +532,13 @@ public:
-    
+
+    future<std::tuple<pid_t, int, int>>
+    spawn(std::string_view pathname,
```
Is there any specific reason why this method should sit on reactor class?

tchaikov replied:
because it needs the access to the `reactor::_thread_pool` for performing the blocking syscalls.

avikivity replied:
Please add a helper in seastar.hh so people don't need to include reactor.hh.

tchaikov replied:
i forgot to address this one. will do so tomorrow.

tchaikov replied:
added.

xemul commented on src/util/process.cc:
```diff
@@ -0,0 +1,83 @@
+    });
+}
+
+future<process> spawn_process(std::string_view pathname,
```
Please, keep it in `experimental` namespace so that we're not doomed to backward compatibility right at once

tchaikov replied:
sure. will do.

xemul commented on src/core/reactor.cc:
```diff
@@ -1921,6 +1924,114 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
+            ret.throw_if_error();
+            return make_ready_future();
+        });
+        return when_all_succeed(std::move(create_cout_pipe),
```
It looks like either pipe would leak if the creation of its peer fails

tchaikov replied:
yes. will fix the leak in the next revision.

xemul commented on src/core/reactor.cc:
```diff
@@ -1921,6 +1924,114 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
+
+future<int> reactor::waitpid(pid_t pid) {
+    return _thread_pool->submit<syscall_result<int>>([pid] {
+        return wrap_syscall<int>(syscall(SYS_pidfd_open, pid, O_NONBLOCK));
```
I guess it's too late to call pidfd_open() already. If it fails for whatever reason, the process is left unwaited-for

tchaikov replied:
true. will fix it.

xemul commented on src/core/reactor.cc:
```diff
@@ -1921,6 +1924,114 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
+                                                       const_cast<char* const *>(argvp.data()),
+                                                       const_cast<char* const *>(envp.data())));
+            });
+        }).then([&child_pid, &cout_pipe, &cerr_pipe, &actions, &attr] (syscall_result<int> ret) {
```
If posix_spawn fails, this .then() chain is not called, the actions and attrs are left not destroyed -> probably leak both

tchaikov replied:
agreed. the error handling basically is missing here. will add it in the next revision.

xemul commented on src/core/reactor.cc:
```diff
@@ -1921,6 +1924,114 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
 
+future<std::tuple<pid_t, int, int>>
+reactor::spawn(std::string_view pathname,
+               const std::vector<std::string_view>& argv,
```
You copy argv and env inside do_with anyway, why not making the API function accept argv/env vectors by value from the very beginning?

tchaikov replied:
okay. changed to 
```c++
  reactor::spawn(std::string_view pathname,
                 std::vector<sstring> argv,
                 std::vector<sstring> env)
```

xemul commented on include/seastar/util/process.hh:
```diff
@@ -0,0 +1,56 @@
+};
+
+future<process> spawn_process(std::string_view pathhname,
+                              const std::vector<std::string_view>& argv,
```
Why not removing the `pathname` completely and assuming that argv[0] is the binary name itself? Like python's subprocess.Popen does

tchaikov replied:
yes. this would simplify the API. actually, i was struggling when it came to the interface of `spawn_process()` in two perspectives:

1. shall we combine stdout and stderr? namely to `adddup2()` both `STDOUT_FILENO` and `STDERR_FILENO` to a single fd. this would simplify the error handling and interface a lot. but the downside is that there might be users want to separate stderr from stdout.
2. use different `pathname` and `argv[0]`? again, i was thinking of a "niche market", where user might need to use tools like "BusyBox". it is a single monolithic executable which expose different unix utilities to end user by checking `argv[0]`.

i might be overthinking the requirement. but at least, these are the reason why the interface looks this way.

tchaikov replied:
i think the new API (partially) addresses this usability issue.

avikivity replied:
I think we shouldn't try to improve posix_spawn, just adapt it to Seastar.

tchaikov commented:
changelog

- move the `spawn_process()` and `process` class into `seastar::experimental` namespace. as these interfaces are subject to change in future.
- handle the errors and do proper cleanup.
- use periodically `waitpid()` call if `pidfd_open()` is not available
- explain why `pidfd_open()` is used in the commit message.

tchaikov commented:
if we believe that the `pathname` and `argv[0]` are always identical, and the use case of BusyBox does not exist, i will drop `pathname` parameter.

if we believe that `stderr` should be redirect to `stdout`, i will update the pipe settings accordingly. not sure if we should allow developer to programmatically customize this behavior with `spawn()` API, though. maybe that'd be unnecessarily too flexible..

tchaikov commented on src/core/reactor.cc:
```diff
@@ -1921,6 +1925,143 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
+        }).then([&child_pid, &cout_pipe, &cerr_pipe] (syscall_result<int> ret) {
+            ::close(cout_pipe[1]);
+            ::close(cerr_pipe[1]);
+            ret.throw_if_error();
```
`cout_pipe[0]` and `cout_pipe[1]` are leaked if `posix_spawn()` fails.

avikivity commented on include/seastar/util/process.hh:
```diff
@@ -0,0 +1,61 @@
+    }
+    input_stream<char>& stderr() {
+        return _stderr;
+    }
```
What about stdin?

I suppose it can be left to a future iteration, no compatibility problem.

tchaikov replied:
added the stdin support.

avikivity commented on include/seastar/util/process.hh:
```diff
@@ -0,0 +1,61 @@
+        T value;
+    };
+    using exit_status = tagged_type<int, struct exit_status_tag>;
+    using term_sig = tagged_type<int, struct term_sig_tag>;
```
How about struct normal_exit { int exit_code }, struct abnormal_exit { int terminating_signal }? To avoid random templates in high level code.

tchaikov replied:
sure. now use plain `strut` instead. but changed the struct name to `wait_exited` and `wait_signaled` respectively.

avikivity commented on include/seastar/util/process.hh:
```diff
@@ -0,0 +1,61 @@
+future<process> spawn_process(std::string_view pathname,
+                              std::vector<sstring> argv,
+                              std::vector<sstring> env = {});
+
```
posix_spawn has about a million options. We don't need to support them all (now), but we need to indicate what's supported. Is this the spawnp variant? Or the regular one?

tchaikov replied:
documented in the doxygen comment.

avikivity commented on src/core/reactor.cc:
```diff
@@ -1921,6 +1925,147 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
+            r = ::posix_spawn_file_actions_addclose(&actions, cout_pipe[1]);
+            throw_system_error_on(r != 0, "posix_spawn_file_actions_addclose");
+            r = ::posix_spawn_file_actions_addclose(&actions, cerr_pipe[1]);
+            throw_system_error_on(r != 0, "posix_spawn_file_actions_addclose");
```
Those are fun!

tchaikov replied:
indeed =)

avikivity commented on src/core/reactor.cc:
```diff
@@ -1921,6 +1925,147 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
+                std::vector<const char*> envp;
+                std::transform(env.cbegin(), env.cend(), std::back_inserter(envp),
+                               [](auto& s) { return s.c_str(); });
+                envp.push_back(nullptr);
```
I suggest to do these transforms outside the worker thread. It used not to support allocations at all, now it does, but it's still a weaker environment.

tchaikov replied:
moved to the outside of the worker thread.

avikivity commented on src/core/reactor.cc:
```diff
@@ -1921,6 +1925,147 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
+    return _thread_pool->submit<syscall_result<int>>([pid] {
+        return wrap_syscall<int>(syscall(SYS_pidfd_open, pid, O_NONBLOCK));
+    }).then([pid, this] (syscall_result<int> pidfd) {
+        if (pidfd.result == -1) {
```
Check for ENOSYS too, could be a different failure.

avikivity replied:
Ah, you do check.

avikivity commented on src/core/reactor.cc:
```diff
@@ -1921,6 +1925,147 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
+                        return wrap_syscall<pid_t>(::waitpid(pid, &wstatus, WNOHANG));
+                    }).then([&wstatus] (syscall_result<pid_t> ret) {
+                        if (ret.result == 0) {
+                            return ::seastar::sleep(wait_pid_poll_period).then([] {
```
This could start lower and increase, so that tiny forks are fast. But not needed in the first iteration.

tchaikov replied:
added an incremental backoff timeout machinery.

avikivity commented on src/core/reactor.cc:
```diff
@@ -1921,6 +1925,147 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
+                return pidfd.readable().then([pid, &wstatus, this] {
+                    return _thread_pool->submit<syscall_result<pid_t>>([pid, &wstatus] {
+                        return wrap_syscall<pid_t>(::waitpid(pid, &wstatus, WNOHANG));
+                    });
```
Do we miss a close()?

tchaikov replied:
sorry, could you please be more specific? if what you are referring is the fd returned by `pidfd_open()`, then it should be closed by the destructor of `file_desc`. the `do_with()` helper should take care of the life cycle of the `with`ed variables.

avikivity replied:
ok

avikivity commented on tests/unit/spawn_test.cc:
```diff
@@ -0,0 +1,82 @@
+                return p.wait().discard_result();
+            });
+        });
+    });
```
Please add a case for /non/existent/path.

tchaikov replied:
sure. added.

avikivity commented:
Looks good overall.

tchaikov commented:
changelog

- added a test case testing `/non/existent/path`
- added an incremental backoff timeout machinery when `waitpid()` returns 0
- moved argv and env preparation into the outside of the worker thread.
- documented all the public interfaces.
- do not use template for creating the typed values for holding the wait status
- added stdin support
- mark the ctor of `process` private using an extra tag parameter.
- moved the `spawn_process()` function into `process` class.

tchaikov commented on include/seastar/util/process.hh:
```diff
@@ -0,0 +1,61 @@
+
+class process {
+public:
+    process(pid_t pid, int stdout, int stderr);
```
ack. added a private tag struct as its first parameter.

xemul commented on src/core/reactor.cc:
```diff
@@ -1921,6 +1925,191 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
+            std::vector<const char*> argvp;
+            std::transform(argv.cbegin(), argv.cend(), std::back_inserter(argvp),
+                           [](auto& s) { return s.c_str(); });
+            argvp.push_back(nullptr);
```
What if it vector::push_back() throws ENOMEM, will pipe descriptors leak?

tchaikov replied:
will address this one in the next revision.

tchaikov replied:
fixed.

xemul commented on src/core/reactor.cc:
```diff
@@ -1921,6 +1925,191 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
+                   std::move(argv),
+                   std::move(env),
+                   [this](auto& child_pid, auto& cin_pipe, auto& cout_pipe, auto& cerr_pipe, auto& pathname, auto& actions, auto& attr, auto& argv, auto& env) {
+        auto create_cin_pipe = _thread_pool->submit<syscall_result<int>>([&cin_pipe] {
```
A suggestion (for next PR) -- it's worth having another API call (in namespace experimental). This place should become a one-line then.

tchaikov replied:
good point. will do so right after this change lands.

tchaikov replied:
#1300

xemul commented on include/seastar/util/process.hh:
```diff
@@ -0,0 +1,98 @@
+class process {
+    struct create_tag {};
+public:
+    process(create_tag, pid_t pid, int stdin, int stdout, int stderr);
```
The constructor can be made private, it's only called from places that have access to create_tag

tchaikov replied:
no. i don't think so. it's the lambda which creates calls the ctor. and it'd be difficult to mark the lambda `process::spawn()` overloads `friend` without using an proxy method, which can be marked as a friend though.

tchaikov commented on tests/unit/spawn_test.cc:
```diff
@@ -0,0 +1,113 @@
+            }).then([&p] {
+                return p.stdout().read_exactly(text.size());
+            }).then([] (temporary_buffer<char> echo) {
+                BOOST_CHECK(echo == echo.copy_of(text));
```
could have used `BOOST_CHECK_EQ()` if we have `operator<<` for `temporary_buffer<char>`. will add it later.

tchaikov commented on tests/unit/spawn_test.cc:
```diff
@@ -0,0 +1,113 @@
+}
+
+SEASTAR_TEST_CASE(test_spawn_program_does_not_exist) {
+    const char* not_exist_cmd = "i/dont/exist";
```
s/not_exists/nonexistent_cmd/. better English.

tchaikov commented on tests/unit/spawn_test.cc:
```diff
@@ -0,0 +1,113 @@
+SEASTAR_TEST_CASE(test_spawn_input) {
+    const char* cat_cmd = "/usr/bin/cat";
+    static const sstring text = "hello world\n";
+    return process::spawn(cat_cmd, {cat_cmd}).then([] (auto process) {
```
```suggestion
    return process::spawn(cat_cmd).then([] (auto process) {
```

tchaikov commented on tests/unit/spawn_test.cc:
```diff
@@ -0,0 +1,113 @@
+
+SEASTAR_TEST_CASE(test_spawn_success) {
+    const char* true_cmd = "/usr/bin/true";
+    return process::spawn(true_cmd, {true_cmd}).then([] (auto process) {
```
```suggestion
    return process::spawn(true_cmd).then([] (auto process) {
```

tchaikov commented on tests/unit/spawn_test.cc:
```diff
@@ -0,0 +1,113 @@
+
+SEASTAR_TEST_CASE(test_spawn_failure) {
+    const char* false_cmd = "/usr/bin/false";
+    return process::spawn(false_cmd, {false_cmd}).then([] (auto process) {
```
```suggestion
    return process::spawn(false_cmd).then([] (auto process) {
```

tchaikov commented on tests/unit/spawn_test.cc:
```diff
@@ -0,0 +1,113 @@
+
+SEASTAR_TEST_CASE(test_spawn_program_does_not_exist) {
+    const char* not_exist_cmd = "i/dont/exist";
+    return process::spawn(not_exist_cmd, {not_exist_cmd}).then_wrapped([] (future<process> fut) {
```
```suggestion
    return process::spawn(not_exist_cmd).then_wrapped([] (future<process> fut) {
```

tchaikov commented on tests/unit/spawn_test.cc:
```diff
@@ -0,0 +1,113 @@
+#include <seastar/core/temporary_buffer.hh>
+#include <seastar/testing/test_case.hh>
+#include <seastar/testing/thread_test_case.hh>
+#include <string>
```
please remove this include as well.

tchaikov commented on include/seastar/util/process.hh:
```diff
@@ -0,0 +1,98 @@
+    using wait_status = std::variant<wait_exited, wait_signaled>;
+    /// Wait until the child process exits or terminates
+    ///
+    /// \returns the exist status
```
typo

tchaikov commented:
changelog

- fix a typo in comment
- place the comma before the initializer in the initializer list. for better consistency.
- remove unused `include <...>` in `tests/unit/spawn_test.cc`.
- use the single-argument `process::spawn()` overload in tests. simpler this way.
- s/not_exists/nonexistent_cmd/ in the variable name in a test
- handle exceptions thrown when preparing `posix_spawn()` argument, and do necessary cleanups.
- move `posix_spawn_file_actions_destroy()` cleanups up, as `_thread_pool->submit<..>()` does not really throw, while the `::posix_spawn*()` calls and `vector<>::push_back()` could.
- s/throw_system_error_on/throw_pthread_error/. because, unlike the syscalls, these POSIX APIs return error with the return value directly.

avikivity commented on include/seastar/core/seastar.hh:
```diff
@@ -384,4 +384,29 @@ future<uint64_t> fs_avail(std::string_view name) noexcept;
+future<std::tuple<pid_t, int, int, int>>
+spawn(std::string_view pathname,
+      std::vector<sstring> argv,
+      std::vector<sstring> env = {});
```
Idea: change the signature to `span(std::filesystem::path path, spawn_parameters params)`. This way when we later add more parameters from posix_spawn() we don't need to add more and more overloads.

Also, shouldn't it return a seastar::process?

tchaikov replied:
love it! actually, i wanted to add
```c++
bool stdin;
bool stdout;
bool stderr;
```
parameters allowing user to enable/disable writing/reading to/from the spawned process. but on the other hand, i don't want to have an overcomplicated API like Python's `subprocess.Popen()` constructor. this setting is a nice compromise, i think.

tchaikov commented:
changelog

- define `spawn_parameters` struct, so it is more extensible. as we are likely to add more options into this struct in future.
- move the spawn helper into `seastar::experimental` namespace
- add doxygen document for the new `subprocess-module` group
- change the spawn helper in `seastar.hh`, so it 
  - returns a `future<process>`
  - accepts `const std::filesystem::path&` and
  - accepts `spawn_parameters params`
- let `process::spawn()` accept `std::filesystem::path pathname` and `spawn_parameters params`.

avikivity commented on include/seastar/core/seastar.hh:
```diff
@@ -384,4 +385,26 @@ future<uint64_t> fs_avail(std::string_view name) noexcept;
+/// the subprocess is spawned with \c posix_spawn() system call, so the
+/// pathname should be relative or absolute path of the executable.
+future<process> spawn_process(const std::filesystem::path& pathname,
+                              spawn_parameters params = {});
```
spawn_parameters isn't defined yet.

tchaikov replied:
it's defined by the included `seastar/util/process.hh`. the purpose of including it is to make `seastar.hh` self-contained. but yes, this is inconsistent with how we, for instance, implement the `connected_socket` related API. i replaced `#include <seastar/util/process.hh>` with forward declarations. but i still think it'd be better to be self-contained, so user does not need to include other headers when using API's declared by `seastar.h`. if this is desirable, i will create another change to address this.

avikivity replied:
Missed it. It's more accurate to have spawn_parameters.hh so that seastar.hh is kept thin, but not worth it for now.

tchaikov replied:
but for accessing `process` one still needs to include `seastar/util/process.hh`. that's exactly what https://github.com/scylladb/seastar/blob/e9511e1b9e9596ea167e9a4001104c83c2c5a032/tests/unit/spawn_test.cc#L22-L25 is doing.

avikivity commented on include/seastar/util/process.hh:
```diff
@@ -0,0 +1,112 @@
+        : argv(argv.begin(), argv.end()) {}
+    spawn_parameters(std::initializer_list<std::string_view> argv,
+                     std::initializer_list<std::pair<std::string_view, std::string_view>> envs)
+        : argv(argv.begin(), argv.end())
```
Parameters struct are best left without constructors, so people can use designated initializers.

tchaikov replied:
i see. dropped.

avikivity commented on include/seastar/util/process.hh:
```diff
@@ -0,0 +1,112 @@
+                       [](const auto& e) {
+                           return fmt::format("{}={}", e.first, e.second);
+                       });
+    }
```
And things like that are best left to the user, or maybe a helper. We should be providing a thin layer and adapt to the async nature of Seastar, not try to change the posix API.

tchaikov replied:
dropped. thought it was also important to have a user friendly API, but yeah, it might diverge from the goal of Seastar, and could open a door of more shim-like APIs.

avikivity commented on include/seastar/util/process.hh:
```diff
@@ -0,0 +1,112 @@
+    ///
+    /// \returns a \c process instance representing the spawned subprocess
+    static future<process> spawn(const std::filesystem::path& pathname,
+                                 spawn_parameters params);
```
Should this be private?

avikivity commented on include/seastar/util/process.hh:
```diff
@@ -0,0 +1,112 @@
+    input_stream<char>& stderr() {
+        return _stderr;
+    }
+    struct wait_exited {
```
Should those move away stdin/stdout? Like connected_socket? I guess it's better to keep the code as is.

tchaikov replied:
instead of constructing these streams in the ctor, the latest revision construct them right in the accessors.

avikivity commented on src/core/reactor.cc:
```diff
@@ -4333,6 +4516,17 @@ void set_idle_cpu_handler(idle_cpu_handler&& handler) {
+    if (params.argv.empty()) {
+        return process::spawn(pathname);
+    } else {
+        return process::spawn(pathname, std::move(params));
```
Don't understand this if. Why not always pass params? They could include more than the environment.

tchaikov replied:
ahh, indeed. i am exposing two APIs in `seastar.hh` now. one with params, one without.

tchaikov commented:
changelog

- drop the constructors for `spawn_parameters`.
- construct streams right in the `process`'s accessors.
- mark `process::spawn()` private
- test `seastar::spawn_process` instead of `process::spawn()` in `spawn_test.cc`
- define two separate API of `spawn_process()`. one with `spawn_parameters`, one without.
- replace `#include <seastar/util/process.hh>` with forward declarations in `seastar.hh` to be more consistent.
